Keep attributes for Modules and Functions

ppetrovicTT opened the pull request:
Hello!

As this is my first time reaching out, I wanted to thank you for this library. At Tenstorrent we started using torchview to better understand external models we're working on, and it helped a lot!

I found that with a small tweak, I could use torchview to list the operators used by a model, with given attributes.
This is an interesting insight, as it can unveil if we have missing operators implementations on our side.

I tried to keep the change the least invasive as possible. And I've added two small tests to help with stability.

Please let me know if this change is interesting for you? It would be great if I could upstream it.

Thank you,
Pavle Petrovic, Tenstorrent

mert-kurttutan commented:
Hi @ppetrovicTT , thanks for the interest, glad it could you. 
Sorry for the late reply. IRL stuff.

There is one type checking error regarding list. It has indexing support starting 3.9. I want to keep 3.8 test a few monts more.

Could you use `from typing import List`?

mert-kurttutan commented:
The other change I suggest is the default behaviour.
As of now, the current behaviour of the library that it collects the graph from running. For the time being I want to preserve this.
So, can you add new parameter to the `draw_graph` function (e.g. collect_attribute) that controls this behaviour.
False -> no collection (i.e. set attribute to None)
True -> collect attributes

It should have default value set to False

Other than this, it is a nice functionality, good to go.

ppetrovicTT commented:
Hello! Thank you for your feedback!
And now I apologize for late iteration, my other threads got higher priority :)

I've followed your advises and fixed mypy errors - as you said it was quite straightforward and good practice.

I've added a flag to draw_graph, but it turned out to be quite invasive, I'm not sure if I like it.
That's why I committed it separately, so that we can decide easier if we want the change or not.

Let me know how it looks like now!

Best,
Pavle

mert-kurttutan commented:
The level of invasiveness is the same as other context_attributes. So, to me, it is not annoying enough to get rid of it.

mert-kurttutan commented:
Can you add `from __future__ import annotations` to the test_attributes.py file?

ppetrovicTT commented:
Thank you very much for merging my changes, @mert-kurttutan! 🎉 

Can't wait for my change to reach the official tag.
Do you know maybe when can we expect that to happen?

Best,
Pavle

mert-kurttutan commented:
It will be done on this Friday at latest.

mert-kurttutan commented:
@ppetrovicTT version 0.2.7 is available on pypi 👍

ppetrovicTT commented:
Amazing! Thank you very much! 🙌 

I overlooked one use-case, so I've created a new PR #128 
I hope you don't mind 🙏
